File: web/src/api/deviceTools.ts
import { get, post, del } from './http'

export type DeviceToolType = 'desktop' | 'browser'

// One instruction of the dynamic MCP program (call / set / return).
export interface DynamicToolStep {
  op: 'call' | 'set' | 'return'
  tool?: string
  args?: Record<string, unknown>
  name?: string
  value?: unknown
  save_as?: string
}

export type ToolRuntime = 'python' | 'powershell' | 'shell'

export interface DynamicToolDefinition {
  name: string
  description: string
  input_schema: Record<string, unknown>
  // 'js' (desktop): server-stored JS run on the device with (args, cap, ctx).
  // 'program' (browser): the call/set/return DSL.
  // 'runtime' (desktop): plain source run by a device runtime (python/powershell/shell).
  code_kind?: 'js' | 'program' | 'runtime'
  code?: DynamicToolStep[]
  js?: string
  runtime?: ToolRuntime
  source?: string
  permissions?: string[]
}

export type ToolStatus = 'active' | 'draft' | 'disabled' | 'archived'

export interface DeviceDynamicTool extends DynamicToolDefinition {
  enabled: boolean
  status: ToolStatus
  revision: string
  updated_at: number
}

export interface DeviceToolsResponse {
  deviceType: DeviceToolType
  tools: DeviceDynamicTool[]
  // Tools a live device of this type currently advertises — offered as call targets.
  availableTools: { name: string; description: string }[]
}

export const listDeviceTools = (deviceType: DeviceToolType) =>
  get<DeviceToolsResponse>('/api/device-tools', {
    query: { device_type: deviceType },
    fallbackError: '设备动态 MCP 工具加载失败',
  })

export const upsertDeviceTool = (
  deviceType: DeviceToolType,
  definition: DynamicToolDefinition,
  enabled = true,
) =>
  post<{ tool: DeviceDynamicTool; pushedToDevices: number }>(
    '/api/device-tools',
    { device_type: deviceType, definition, enabled },
    { fallbackError: '保存动态 MCP 工具失败' },
  )

export const toggleDeviceTool = (deviceType: DeviceToolType, name: string, enabled: boolean) =>
  post<{ tool: DeviceDynamicTool; pushedToDevices: number }>(
    '/api/device-tools/toggle',
    { device_type: deviceType, name, enabled },
    { fallbackError: '切换动态 MCP 工具失败' },
  )

// Approve a draft (status='active') or shelve a tool (draft/disabled/archived).
export const setDeviceToolStatus = (deviceType: DeviceToolType, name: string, status: ToolStatus) =>
  post<{ tool: DeviceDynamicTool; pushedToDevices: number }>(
    '/api/device-tools/status',
    { device_type: deviceType, name, status },
    { fallbackError: '更新工具状态失败' },
  )

export const deleteDeviceTool = (deviceType: DeviceToolType, name: string) =>
  del<{ ok: boolean; pushedToDevices: number }>(
    `/api/device-tools/${encodeURIComponent(name)}`,
    { query: { device_type: deviceType }, fallbackError: '删除动态 MCP 工具失败' },
  )

export interface DeviceToolVersion {
  version_id: number
  name: string
  revision: string
  action: 'upsert' | 'delete' | 'restore'
  actor: 'web' | 'ai'
  ai_config_id: number | null
  description: string
  code_kind: 'js' | 'program' | 'runtime'
  created_at: number
}

export const listDeviceToolVersions = (deviceType: DeviceToolType, name: string) =>
  get<{ deviceType: DeviceToolType; name: string; versions: DeviceToolVersion[] }>(
    '/api/device-tools/versions',
    { query: { device_type: deviceType, name }, fallbackError: '历史版本加载失败' },
  )

export const restoreDeviceToolVersion = (deviceType: DeviceToolType, versionId: number) =>
  post<{ tool: DeviceDynamicTool; pushedToDevices: number }>(
    '/api/device-tools/restore',
    { device_type: deviceType, version_id: versionId },
    { fallbackError: '回滚失败' },
  )

export interface DeviceToolStat {
  tool: string
  total: number
  failures: number
  failure_rate: number
  last_called_at: number
  last_failure_at: number
  last_error: string
}

export interface DeviceToolFailure {
  tool: string
  error: string
  ai_config_id: number | null
  session_id: string
  run_id: string
  message_id: number | null
  created_at: number
}

export const listDeviceToolStats = (deviceType: DeviceToolType) =>
  get<{ deviceType: DeviceToolType; stats: DeviceToolStat[] }>(
    '/api/device-tools/stats',
    { query: { device_type: deviceType }, fallbackError: '调用统计加载失败' },
  )

export const listDeviceToolFailures = (name: string) =>
  get<{ name: string; failures: DeviceToolFailure[] }>(
    '/api/device-tools/failures',
    { query: { name }, fallbackError: '失败记录加载失败' },
  )

// Server-governed permission policy for runtime tools: maps a permission tag to
// allow / confirm / deny. Shipped to devices in the tool-config push.
export type PermissionDecision = 'allow' | 'confirm' | 'deny'

export const getPermissionPolicy = (deviceType: DeviceToolType) =>
  get<{ deviceType: DeviceToolType; policy: Record<string, PermissionDecision>; knownTags: string[] }>(
    '/api/device-tools/permission-policy',
    { query: { device_type: deviceType }, fallbackError: '权限策略加载失败' },
  )

// Which runtimes the user's online devices of this type can actually run.
export const getDeviceRuntimes = (deviceType: DeviceToolType) =>
  get<{ deviceType: DeviceToolType; runtimes: Record<ToolRuntime, boolean> }>(
    '/api/device-tools/runtimes',
    { query: { device_type: deviceType }, fallbackError: '运行时能力加载失败' },
  )

export const setPermissionPolicy = (deviceType: DeviceToolType, policy: Record<string, PermissionDecision>) =>
  post<{ deviceType: DeviceToolType; policy: Record<string, PermissionDecision>; pushedToDevices: number }>(
    '/api/device-tools/permission-policy',
    { device_type: deviceType, policy },
    { fallbackError: '保存权限策略失败' },
  )

File: server/gateway/routers/device_tools.py
"""Device dynamic MCP tools: CRUD for web-authored, device-type-scoped tools.

An operator manages dynamic MCP tools (``call`` / ``set`` / ``return`` JSON
programs) per device type (``desktop`` / ``browser``) from the web console. The
server validates them the same way the device interpreter does, stores them,
and pushes the enabled set to every online device of that type so a tool change
ships without a new client release.
"""

from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Body, Depends, Header, HTTPException, Query
from pydantic import BaseModel, Field
from sqlmodel import Session

from api.database import get_session
from api.device_live import push_device_dynamic_tools
from api.device_presence import online_tool_catalog_for_user
from api.services import device_workspace_tools as dyn
from .auth import get_current_user

router = APIRouter()
PREFIX = "/api/device-tools"


class DynamicToolUpsert(BaseModel):
    device_type: str = Field(..., description="desktop | browser")
    definition: Dict[str, Any] = Field(..., description="{name, description, input_schema, code}")
    enabled: bool = True


class DynamicToolToggle(BaseModel):
    device_type: str
    name: str
    enabled: bool


class DynamicToolStatus(BaseModel):
    device_type: str
    name: str
    status: str = Field(..., description="active | draft | disabled | archived")


class DynamicToolRestore(BaseModel):
    device_type: str
    version_id: int


class PermissionPolicyUpdate(BaseModel):
    device_type: str
    policy: Dict[str, str] = Field(default_factory=dict, description="{permission_tag: allow|confirm|deny}")


def _available_call_targets(user_id: int, device_type: str) -> List[Dict[str, str]]:
    """Tool names the visual editor can offer as ``call`` targets: every tool an
    online device of this type currently advertises (built-ins included), so the
    operator picks from what the device can actually run."""
    targets: Dict[str, str] = {}
    for device in online_tool_catalog_for_user(user_id):
        if str(device.get("device_type") or "") != device_type:
            continue
        for tool in device.get("tools") or []:
            name = str(tool.get("name") or "").strip()
            if name:
                targets.setdefault(name, str(tool.get("description") or "").strip())
    return [{"name": name, "description": targets[name]} for name in sorted(targets)]


@router.get("")
async def list_device_tools(
    device_type: str = Query(...),
    session: Session = Depends(get_session),
    authorization: str = Header(None),
):
    user = get_current_user(authorization, session)
    try:
        dtype = dyn.normalize_device_type(device_type)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    return {
        "deviceType": dtype,
        "tools": dyn.list_tools(user.id, dtype),
        # Tools the operator can target in a ``call`` step (live device catalog).
        "availableTools": _available_call_targets(user.id, dtype),
    }


@router.post("")
async def upsert_device_tool(
    payload: DynamicToolUpsert,
    session: Session = Depends(get_session),
    authorization: str = Header(None),
):
    user = get_current_user(authorization, session)
    try:
        dtype = dyn.normalize_device_type(payload.device_type)
        tool = dyn.upsert_tool(user.id, dtype, payload.definition, enabled=payload.enabled)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    reached = await push_device_dynamic_tools(user.id, dtype)
    return {"tool": tool, "pushedToDevices": reached}


@router.post("/toggle")
async def toggle_device_tool(
    payload: DynamicToolToggle,
    session: Session = Depends(get_session),
    authorization: str = Header(None),
):
    user = get_current_user(authorization, session)
    try:
        dtype = dyn.normalize_device_type(payload.device_type)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    tool = dyn.set_enabled(user.id, dtype, payload.name, payload.enabled)
    if tool is None:
        raise HTTPException(status_code=404, detail="Dynamic MCP tool not found")
    reached = await push_device_dynamic_tools(user.id, dtype)
    return {"tool": tool, "pushedToDevices": reached}


@router.post("/status")
async def set_device_tool_status(
    payload: DynamicToolStatus,
    session: Session = Depends(get_session),
    authorization: str = Header(None),
):
    """Approve (active) or shelve (draft/disabled/archived) a tool. AI-authored
    tools land as draft; this is how an operator promotes them."""
    user = get_current_user(authorization, session)
    try:
        dtype = dyn.normalize_device_type(payload.device_type)
        tool = dyn.set_status(user.id, dtype, payload.name, payload.status)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    if tool is None:
        raise HTTPException(status_code=404, detail="Dynamic MCP tool not found")
    reached = await push_device_dynamic_tools(user.id, dtype)
    return {"tool": tool, "pushedToDevices": reached}


@router.get("/stats")
async def device_tool_stats(
    device_type: str = Query(...),
    session: Session = Depends(get_session),
    authorization: str = Header(None),
):
    user = get_current_user(authorization, session)
    from api.services import mcp_stats

    try:
        dtype = dyn.normalize_device_type(device_type)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    tool_names = [t["name"] for t in dyn.list_tools(user.id, dtype)]
    return {"deviceType": dtype, "stats": mcp_stats.tool_stats(user.id, tool_names)}


@router.get("/failures")
async def device_tool_failures(
    name: str = Query(...),
    session: Session = Depends(get_session),
    authorization: str = Header(None),
):
    user = get_current_user(authorization, session)
    from api.services import mcp_stats

    return {"name": name, "failures": mcp_stats.recent_failures(user.id, name)}


@router.get("/versions")
async def list_device_tool_versions(
    device_type: str = Query(...),
    name: str = Query(...),
    session: Session = Depends(get_session),
    authorization: str = Header(None),
):
    user = get_current_user(authorization, session)
    try:
        dtype = dyn.normalize_device_type(device_type)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    return {"deviceType": dtype, "name": name, "versions": dyn.list_versions(user.id, dtype, name)}


@router.get("/version")
async def get_device_tool_version(
    device_type: str = Query(...),
    version_id: int = Query(...),
    session: Session = Depends(get_session),
    authorization: str = Header(None),
):
    user = get_current_user(authorization, session)
    try:
        dtype = dyn.normalize_device_type(device_type)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    snapshot = dyn.get_version(user.id, dtype, version_id)
    if snapshot is None:
        raise HTTPException(status_code=404, detail="Version not found")
    return {"version": snapshot}


@router.post("/restore")
async def restore_device_tool(
    payload: DynamicToolRestore,
    session: Session = Depends(get_session),
    authorization: str = Header(None),
):
    user = get_current_user(authorization, session)
    try:
        dtype = dyn.normalize_device_type(payload.device_type)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    tool = dyn.restore_version(user.id, dtype, payload.version_id, actor="web")
    if tool is None:
        raise HTTPException(status_code=404, detail="Version not found")
    reached = await push_device_dynamic_tools(user.id, dtype)
    return {"tool": tool, "pushedToDevices": reached}


@router.get("/runtimes")
async def device_runtimes(
    device_type: str = Query("desktop"),
    session: Session = Depends(get_session),
    authorization: str = Header(None),
):
    """Which runtimes (python/powershell/shell) the user's online devices of this
    type can actually execute — drives the 'no online device supports X' hint."""
    user = get_current_user(authorization, session)
    from connector_runtime.dispatch.desktop_device_tools import online_runtimes

    try:
        dtype = dyn.normalize_device_type(device_type)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    return {"deviceType": dtype, "runtimes": online_runtimes(user.id, dtype)}


@router.get("/permission-policy")
async def get_permission_policy(
    device_type: str = Query(...),
    session: Session = Depends(get_session),
    authorization: str = Header(None),
):
    user = get_current_user(authorization, session)
    from api.services import device_permission_policy as policy_svc

    try:
        dtype = dyn.normalize_device_type(device_type)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    return {
        "deviceType": dtype,
        "policy": policy_svc.get_policy(user.id, dtype),
        "knownTags": policy_svc.known_tags(),
    }


@router.post("/permission-policy")
async def set_permission_policy(
    payload: PermissionPolicyUpdate,
    session: Session = Depends(get_session),
    authorization: str = Header(None),
):
    user = get_current_user(authorization, session)
    from api.services import device_permission_policy as policy_svc

    try:
        dtype = dyn.normalize_device_type(payload.device_type)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    policy = policy_svc.set_policy(user.id, dtype, payload.policy)
    # Push so online devices apply the new policy immediately (it rides in the
    # device:tool-config payload).
    reached = await push_device_dynamic_tools(user.id, dtype)
    return {"deviceType": dtype, "policy": policy, "pushedToDevices": reached}


@router.delete("/{name}")
async def delete_device_tool(
    name: str,
    device_type: str = Query(...),
    session: Session = Depends(get_session),
    authorization: str = Header(None),
):
    user = get_current_user(authorization, session)
    try:
        dtype = dyn.normalize_device_type(device_type)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    if not dyn.delete_tool(user.id, dtype, name):
        raise HTTPException(status_code=404, detail="Dynamic MCP tool not found")
    reached = await push_device_dynamic_tools(user.id, dtype)
    return {"ok": True, "pushedToDevices": reached}
